spec/security_alert_handler_spec.rb

```diff
@@ -0,0 +1,130 @@
+require "spec_helper"
+require_relative "../lib/security_alert_handler"
+require "ostruct"
+
+RSpec.describe SecurityAlertHandler do
+  let(:github) { double("Github") }
+  let(:organisation) { "test-org" }
+  let(:repos) { %w[repo1 repo2] }
+
+  subject(:handler) do
+    described_class.new(github, organisation, repos)
+  end
+
+  let(:sample_alerts) do
+    [
+      {
+        repo: "repo1",
+        package_manager: "npm",
+        dependency_name: "lodash",
+        vulnerable_version_ranges: [">=4.0.0", "<4.17.13"],
+        url: "https://github.com/test-org/repo1/security/dependabot/package-lock.json/npm/lodash/open",
+        severity: "high",
+      },
+      {
+        repo: "repo2",
+        package_manager: "npm",
+        dependency_name: "lodash",
+        vulnerable_version_ranges: [">=4.0.0", "<4.17.12"],
+        url: "https://github.com/test-org/repo2/security/dependabot/package-lock.json/npm/lodash/open",
+        severity: "high",
+      },
+    ]
+  end
+
+  let(:api_response) do
+    [
+      {
+        dependency: {
+          package: {
+            ecosystem: "npm",
+            name: "lodash",
+          },
+        },
+        security_vulnerability: {
+          vulnerable_version_range: ">=4.0.0, <4.17.13",
+          severity: "high",
+        },
+        html_url: "https://github.com/test-org/repo1/security/dependabot/package-lock.json/npm/lodash/open",
+      },
+    ]
+  end
+
+  before do
+    allow(github).to receive(:get).with(
+      "https://api.github.com/repos/test-org/repo1/dependabot/alerts",
+      accept: "application/vnd.github+json",
+      state: "open",
+    ).and_return(api_response)
+
+    allow(github).to receive(:get).with(
+      "https://api.github.com/repos/test-org/repo2/dependabot/alerts",
+      accept: "application/vnd.github+json",
+      state: "open",
+    ).and_return(api_response)
+
+    allow(handler).to receive(:fetch_security_alerts).and_return(sample_alerts)
+  end
+
+  describe "#security_alerts_count" do
+    context "when an API error occurs for all repos" do
```

Comment:
Looking at the code (and reading this test name), it sounds like we only return `nil` if ALL repos face an API error. If a single repo API call succeeds, the return value is an array. I.e. we only treat "ALL repos facing API error" as a failure, anything else is considered good enough to report on (and potentially misleading, e.g. if only one repo succeeds, but a dozen other repos' API calls don't succeed and there are security alerts tucked away in them, but we're falsely reporting `[]` and implying there are no security concerns)

I'd prefer if we erred on the side of caution and treat a single API failure as an overall failure, i.e. return `nil` if all but one repo API call succeeds. Did you find API calls failing often (does this code hit GitHub's rate limits) or is it just a theoretical issue you've not really come across?

Reply:
I agree that this isn't great, I'm changing it for a better approach. The reason I added this was that some teams have not given the Seal enough permission to see these alerts, so if they don't set `security_alerts` to `false`, the API returns errors and they get a message saying they have 0 alerts, which would be misleading. But as you pointed out, this approach isn't good either, so I'm refactoring this part of the code.

Reply:
Thanks, looks good!